celery/concurrency/asynpool.py

```diff
@@ -464,14 +465,56 @@ def _track_child_process(self, proc, hub):
             # as once the original fd is closed we cannot unregister
             # the fd from epoll(7) anymore, causing a 100% CPU poll loop.
             fd = proc._sentinel_poll = os.dup(proc._popen.sentinel)
-        hub.add_reader(fd, self._event_process_exit, hub, proc)
+        # Safely call hub.add_reader for the determined fd
+        self.iterate_file_descriptors_safely(
+            [fd], None, hub.add_reader,
+            self._event_process_exit, hub, proc)
 
     def _untrack_child_process(self, proc, hub):
         if proc._sentinel_poll is not None:
             fd, proc._sentinel_poll = proc._sentinel_poll, None
             hub.remove(fd)
             os.close(fd)
 
+    def iterate_file_descriptors_safely(self, fds_iter, managed_list,
+                                        hub_method, *args, **kwargs):
+        """Apply hub method to fds in iter, remove from list if failure.
+
+        Some file descriptors may become stale through OS reasons
+        or possibly other reasons, so safely manage our lists of FDs.
+        :param fds_iter: the file descriptors to iterate and apply hub_method
+        :param managed_list: data source to remove FD if it renders OSError
+        :param hub_method: the method to call with with each fd and kwargs
+        :*args to pass through to the hub_method;
+        with a special syntax string '*fd*' represents a substitution
+        for the current fd object in the iteration (for some callers).
+        :**kwargs to pass through to the hub method (no substitutions needed)
+        """
+        def _meta_fd_argument_maker():
+            # uses the current iterations value for fd
+            call_args = args
+            if "*fd*" in call_args:
+                call_args = [fd if arg == "*fd*" else arg for arg in args]
+            return call_args
+        # Track stale FDs for cleanup possibility
+        stale_fds = []
+        for fd in fds_iter:
+            # Handle using the correct arguments to the hub method
+            hub_args, hub_kwargs = args, kwargs  # Default calling pattern
+            if "*fd*" in hub_args:
+                hub_args = _meta_fd_argument_maker()
+            try:  # Call the hub method
+                hub_method(fd, *hub_args, **hub_kwargs)
+            except (OSError, FileNotFoundError) as e:
+                logger.warning(
+                    "Encountered OSError when accessing fd %s ",
+                    fd, exc_info=True)
+                stale_fds.append(fd)  # take note of stale fd
+        # Remove now defunct fds from the managed list
+        if managed_list:
+            for fd in stale_fds:
+                managed_list.pop(fd, None)
```

Comment:
i am currently trying this out. it did not solve my problem OOB, but i am trying if it works, when the `epoll.unregister` is not swallowing all kinds of exceptions. `managed_list` is a `set`, when being called below with `all_inqueues`, which results in a `TypeError`.

Reply:
nvm, you already covered this.

Reply:
Yeah sorry about that -- sent you the link before I added any tests and the existing tests really didn't cover AsyncPool much.